lib/Canonicalize/EmbedStatementComments.cpp

```diff
@@ -0,0 +1,163 @@
+                                                 bool,
+                                                 StringLiteral,
+                                                 StringLiteral>(M.getContext());
+    LLVMIRComment = M.getOrInsertFunction("comment", &FT);
+    auto &Callee = *llvm::cast<llvm::Function>(LLVMIRComment->getCallee());
+    Callee.addFnAttr(llvm::Attribute::NoUnwind);
+    Callee.addFnAttr(llvm::Attribute::WillReturn);
+    Callee.addFnAttr(llvm::Attribute::NoMerge);
+    Callee.setDoesNotAccessMemory();
+    FunctionTags::Comment.addTo(&Callee);
+
+    return LLVMIRComment.value();
+  }
+
+public:
+  inline std::array<pipeline::ContractGroup, 1> getContract() const {
+    return { pipeline::ContractGroup(revng::kinds::StackAccessesSegregated,
+                                     0,
+                                     revng::kinds::StackAccessesSegregated) };
+  }
+
+  void run(pipeline::ExecutionContext &Context,
+           pipeline::LLVMContainer &ModuleContainer) {
+    llvm::Module &M = ModuleContainer.getModule();
+    llvm::IRBuilder<> B(M.getContext());
+
+    using TaggedFunctionKind = revng::kinds::TaggedFunctionKind;
+    llvm::StringRef ContainerName = ModuleContainer.name();
+
+    for (auto &&[ModelFunction, LLVMFunction] :
+         TaggedFunctionKind::getFunctionsAndCommit(Context, M, ContainerName)) {
+      if (ModelFunction->Comments().empty())
+        continue;
+
+      using MapT = yield::CommentPlacementHelper<llvm::BasicBlock *>;
+      MapT CM(*ModelFunction, *LLVMFunction);
+
+      auto EmitAComment = [this, &B, &M](llvm::Instruction *Where,
+                                         const MapT::CommentView &Comment,
+                                         llvm::StringRef EmittedLocation) {
+        B.SetInsertPoint(Where);
+
+        std::array<llvm::Value *, 4> Arguments = {
+          llvm::ConstantInt::get(llvm::Type::getInt64Ty(B.getContext()),
+                                 Comment.CommentIndex),
+          llvm::ConstantInt::get(llvm::Type::getInt8Ty(B.getContext()),
+                                 Comment.LocationMatchesExactly),
+          getUniqueString(&M, addressesToString(*Comment.ExpectedLocation)),
+          getUniqueString(&M, EmittedLocation)
+        };
+
+        auto *Call = B.CreateCall(llvmIRComment(M), Arguments);
+        Call->copyMetadata(*Where);
+      };
+
+      llvm::SmallVector<llvm::Value *, 8> Argumentss;
+      llvm::BasicBlock &EntryBlock = LLVMFunction->getEntryBlock();
+      for (llvm::BasicBlock *Node : llvm::depth_first(&EntryBlock)) {
+        using Trait = yield::StatementTraits<llvm::BasicBlock *>;
+        for (llvm::Instruction *I : Trait::getStatements(Node))
+          for (const MapT::CommentView &Comment : CM.getComments(I))
+            EmitAComment(I, Comment, addressesToString(Trait::getAddresses(I)));
+      }
+
+      for (const MapT::CommentView &Comment :
+           std::views::reverse(CM.getHomelessComments())) {
```

Comment:
Why reverse?

Reply:
I don't remember from the top of my head, I'll look into it.

Reply:
Ah, this was an artifact of a _very_ old design - because of how homeless comments were gathered initially they needed reversed. Not needed anymore, Good catch!